extmod/network_ppp: Add stream config parameter.

DvdGiessen opened the pull request:
This makes the stream that the PPP object wraps, which is normally only set once via the constructor, accessible and configurable via the `.config()` method.

It also allows the stream to be set to `None`, which essentially stops all PPP communication without disconnecting the session.

This allows retrieving the stream, and more interesting, replacing it on-the-fly to suspend it, for example to send AT commands to a modem without completely disconnecting and re-establishing the PPP connection:

```py
uart = ppp.config('stream')
ppp.config(stream=None)
uart.write(b'+++')
# do some AT commands
uart.write(b'ATO\r\n')
ppp.config(stream=uart)
```

Any attempted communication by PPP while the UART is not connected will register as simple packet loss to the LwIP stack because we return 0 for any write calls, and protocols like TCP will then automatically handle retrying.

EDIT: Reworked to also support the generic network.PPP implementation alongside the ESP32 one.

DvdGiessen committed:
extmod/network_ppp: Add stream config parameter.

This makes the stream that the PPP object wraps, which is normally only set
once via the constructor, accessible and configurable via the
`ppp.config()` method.

Signed-off-by: Daniël van de Giessen <daniel@dvdgiessen.nl>

dpgeorge commented:
This is neat, but could there be race conditions?  Eg in your example code above, what if the PPP driver writes data out between the last two lines, won't that data then get lost (never sent to the UART)?

Maybe there needs to be a new method added to PPP which allows you to send AT commands?

Are there any other uses for getting/setting the stream?  You should already know the stream from when you constructed the PPP object.

winnie-xzh commented:
This solution is great, but I see compilation error messages that are simply not accessible to github. When is the next automatic compilation, or can someone manually make it compile again?

DvdGiessen commented:
> Maybe there needs to be a new method added to PPP which allows you to send AT commands?

Possible, but I'm not sure whether that is the way to go. The PPP class and protocol are not aware of what AT is or that it's available as an alternate mode; I think it is OK to keep that separate (after all, you could also use PPP on a connection that has no AT command support; for example some modems have multiple UARTs dedicating one to AT commands and another to the data connections).

Having a more expansive implementation of AT commands, protocol switching, etc in Python that uses these lower-level primitives (what is what I'm doing currently) seems like a good trade-off to keep MicroPython lightweight.

> Are there any other uses for getting/setting the stream? You should already know the stream from when you constructed the PPP object.

Correct, retrieving the stream doesn't really have much direct benefits, you can also store it separately. Retrieving is there for parity with the setter. The main user case for setting it would be changing what PPP communicates with. In my example for suspending it to do AT, but it could also for inspecting the contents, moving it to another physical connection, etc.

> This is neat, but could there be race conditions? Eg in your example code above, what if the PPP driver writes data out between the last two lines, won't that data then get lost (never sent to the UART)?

Ah, yes, there's a slight race condition in my example where some bytes can be lost. I wrote it that way since this way any lost bytes appears as simple packet/byte loss to PPP, which the higher-level protocols can deal with easily. The alternative is to swap those two lines; then no bytes would be lost but (if PPP is actively writing while we write the buffer contents) bytes would be written out of order, and I didn't dig deep enough into how PPP works to see if that could have unintended side-effects.

The other alternative is not replacing the stream but starting in the first place with an custom stream implementation, which then does locking. But that has a few extra downsides things that need looking into, or, to be more precise, there's currently things missing in the PPP implementation that I didn't fix in this MR:

Right now the PPP thread doesn't hold the GIL while interacting with its stream. That is a little bit dangerous, although in practise it (evidently) works fine because usually people directly use PPP with with an UART object. Interfacing with a BytesIO is riskier already (because it is used from a non-uPy thread, reallocation makes it crash). I didn't try making a custom stream that implements the switch in Python; I suppose it could work but it would require a lot more care and understanding of the internals to not accidentally trip up on this GIL-less operation.

DvdGiessen committed:
extmod/network_ppp: Allow stream=None to suspend PPP.

This allows the stream to be set to `None`, which essentially stops all PPP
communication without disconnecting the session.

This allows replacing the stream on-the-fly to suspend it, for example to
send AT commands to a modem without completely disconnecting and
re-establishing the PPP connection:

    uart = ppp.config('stream')
    ppp.config(stream=None)
    uart.write(b'+++')
    # do some AT commands
    uart.write(b'ATO\r\n')
    ppp.config(stream=uart)

Any attempted communication by PPP while the stream is not connected will
register as simple packet loss to the LwIP stack because we return 0 for
any write calls, and protocols like TCP will then automatically handle
retrying.

Signed-off-by: Daniël van de Giessen <daniel@dvdgiessen.nl>

DvdGiessen commented:
I've experimented with it a bit more.

### Attempt 1: Allow PPP callbacks to properly run MicroPython code

My first attempt was trying to set up a full working MicroPython state, so I could implement a `SuspendableStream` class in MicroPython that allows me to suspend PPP without needing to swap out the stream.

<details>
<summary>Patch fully initialising MPy state on the PPP threads</summary>

This uses https://github.com/micropython/micropython/pull/13774 to help set up the uPy thread state.

```diff
diff --git a/ports/esp32/network_ppp.c b/ports/esp32/network_ppp.c
index 8d5c3cb88a..62be4e2f02 100644
--- a/ports/esp32/network_ppp.c
+++ b/ports/esp32/network_ppp.c
@@ -46,6 +46,7 @@
 #if defined(CONFIG_ESP_NETIF_TCPIP_LWIP) && defined(CONFIG_LWIP_PPP_SUPPORT)
 
 #define PPP_CLOSE_TIMEOUT_MS (4000)
+#define PPP_STACK_SIZE (2 * 1024)
 
 typedef struct _ppp_if_obj_t {
     mp_obj_base_t base;
@@ -108,14 +109,54 @@ static u32_t ppp_output_callback(ppp_pcb *pcb, u8_t *data, u32_t len, void *ctx)
         return 0;
     }
 
-    int err;
-    return mp_stream_rw(stream, data, len, &err, MP_STREAM_RW_WRITE);
+    // This callback is called from lwIP's tcpip_thread, which doesn't have
+    // MicroPython state set up. We should set it up it every time, because
+    // we don't know if the tcpip_thread may have been recreated in between
+    mp_state_thread_t *ts_orig = mp_thread_get_state();
+    mp_state_thread_t ts;
+    if (ts_orig == NULL) {
+        mp_thread_init_state(
+            &ts,
+            PPP_STACK_SIZE,
+            mp_state_ctx.thread.dict_locals,
+            mp_state_ctx.thread.dict_globals
+            );
+        // TODO: This causes deadlocks
+        MP_THREAD_GIL_ENTER();
+    }
+
+    nlr_buf_t nlr;
+    if (nlr_push(&nlr) == 0) {
+        int err;
+        len = mp_stream_rw(stream, data, len, &err, MP_STREAM_RW_WRITE);
+        nlr_pop();
+    } else {
+        mp_printf(MICROPY_ERROR_PRINTER, "Unhandled exception writing to PPP stream\n");
+        mp_obj_print_exception(MICROPY_ERROR_PRINTER, MP_OBJ_FROM_PTR(nlr.ret_val));
+        len = 0;
+        self->stream = MP_OBJ_NULL;
+    }
+
+    if (ts_orig == NULL) {
+        MP_THREAD_GIL_EXIT();
+        mp_thread_set_state(ts_orig);
+    }
+
+    return len;
 }
 
 static void pppos_client_task(void *self_in) {
     ppp_if_obj_t *self = (ppp_if_obj_t *)self_in;
     uint8_t buf[256];
 
+    // Set up thread state so we can safely call into MicroPython
+    mp_state_thread_t ts;
+    mp_thread_init_state(
+        &ts,
+        PPP_STACK_SIZE,
+        mp_state_ctx.thread.dict_locals,
+        mp_state_ctx.thread.dict_globals
+        );
     int len = 0;
     while (ulTaskNotifyTake(pdTRUE, len <= 0) == 0) {
         mp_obj_t stream = self->stream;
@@ -123,7 +164,19 @@ static void pppos_client_task(void *self_in) {
             len = 0;
         } else {
             int err;
-            len = mp_stream_rw(stream, buf, sizeof(buf), &err, 0);
+            // TODO: This causes deadlocks
+            MP_THREAD_GIL_ENTER();
+            nlr_buf_t nlr;
+            if (nlr_push(&nlr) == 0) {
+                len = mp_stream_rw(stream, buf, sizeof(buf), &err, 0);
+                nlr_pop();
+            } else {
+                mp_printf(MICROPY_ERROR_PRINTER, "Unhandled exception reading from PPP stream\n");
+                mp_obj_print_exception(MICROPY_ERROR_PRINTER, MP_OBJ_FROM_PTR(nlr.ret_val));
+                len = 0;
+                self->stream = MP_OBJ_NULL;
+            }
+            MP_THREAD_GIL_EXIT();
             if (len > 0) {
                 pppos_input_tcpip(self->pcb, (u8_t *)buf, len);
             }
@@ -145,7 +198,9 @@ STATIC mp_obj_t ppp_active(size_t n_args, const mp_obj_t *args) {
                 return mp_const_true;
             }
 
+            MP_THREAD_GIL_EXIT();
             self->pcb = pppapi_pppos_create(&self->pppif, ppp_output_callback, ppp_status_cb, self);
+            MP_THREAD_GIL_ENTER();
 
             if (self->pcb == NULL) {
                 mp_raise_msg(&mp_type_RuntimeError, MP_ERROR_TEXT("init failed"));
@@ -158,7 +213,9 @@ STATIC mp_obj_t ppp_active(size_t n_args, const mp_obj_t *args) {
 
             if (self->client_task_handle != NULL) { // is connecting or connected?
                 // Wait for PPPERR_USER, with timeout
+                MP_THREAD_GIL_EXIT();
                 pppapi_close(self->pcb, 0);
+                MP_THREAD_GIL_ENTER();
                 uint32_t t0 = mp_hal_ticks_ms();
                 while (!self->clean_close && mp_hal_ticks_ms() - t0 < PPP_CLOSE_TIMEOUT_MS) {
                     mp_hal_delay_ms(10);
@@ -173,7 +230,9 @@ STATIC mp_obj_t ppp_active(size_t n_args, const mp_obj_t *args) {
             }
 
             // Release PPP
+            MP_THREAD_GIL_EXIT();
             pppapi_free(self->pcb);
+            MP_THREAD_GIL_ENTER();
             self->pcb = NULL;
             self->active = false;
             self->connected = false;
@@ -217,19 +276,28 @@ STATIC mp_obj_t ppp_connect_py(size_t n_args, const mp_obj_t *args, mp_map_t *kw
     if (parsed_args[ARG_authmode].u_int != PPPAUTHTYPE_NONE) {
         const char *username_str = mp_obj_str_get_str(parsed_args[ARG_username].u_obj);
         const char *password_str = mp_obj_str_get_str(parsed_args[ARG_password].u_obj);
+        MP_THREAD_GIL_EXIT();
         pppapi_set_auth(self->pcb, parsed_args[ARG_authmode].u_int, username_str, password_str);
+        MP_THREAD_GIL_ENTER();
     }
-    if (pppapi_set_default(self->pcb) != ESP_OK) {
+
+    MP_THREAD_GIL_EXIT();
+    err_t ret = pppapi_set_default(self->pcb);
+    MP_THREAD_GIL_ENTER();
+    if (ret != ESP_OK) {
         mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("set default failed"));
     }
 
     ppp_set_usepeerdns(self->pcb, true);
 
-    if (pppapi_connect(self->pcb, 0) != ESP_OK) {
+    MP_THREAD_GIL_EXIT();
+    ret = pppapi_connect(self->pcb, 0);
+    MP_THREAD_GIL_ENTER();
+    if (ret != ESP_OK) {
         mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("connect failed"));
     }
 
-    if (xTaskCreatePinnedToCore(pppos_client_task, "ppp", 2048, self, 1, (TaskHandle_t *)&self->client_task_handle, MP_TASK_COREID) != pdPASS) {
+    if (xTaskCreatePinnedToCore(pppos_client_task, "ppp", PPP_STACK_SIZE / sizeof(StackType_t), self, 1, (TaskHandle_t *)&self->client_task_handle, MP_TASK_COREID) != pdPASS) {
         mp_raise_msg(&mp_type_RuntimeError, MP_ERROR_TEXT("failed to create worker task"));
     }
 
```

</details>

The problem with this approach is it deadlocks on the GIL. I already added some unlocks in this class itself but that was not sufficient. Specifically, all kinds of code in the rest of the ESP32 calls the lwIP API's (for socket operations etc) without releasing the GIL. These do their work on the lwIP's `tcpip_thread`, which in turn calls the output callback from that thread. So acquiring the GIL there deadlocks the entire board.

I didn't want to go through all those functions and release the GIL there. In some code it already unlocks the GIL but not always (for performance reasons when doing many small lwIP operations).

Perhaps this is still an improvement over the current state, and as a side-effect of the process described above it might work for some code since while we're not technically holding the GIL, we know some other thread is and that thread is waiting on us. But I didn't try this out further.

I read [here](https://github.com/micropython/micropython/issues/4895#issuecomment-513615788) that the ESP32 port could support running without the GIL if some extra locking is implemented in some ISR's. Sounds interesting, but I didn't yet look into it. Perhaps that might also sidestep this issue.

### Attempt 2: Make suspending easier

Based on the logic that data being lost reads like packet loss to the network stack; I added a small feature allows the PPP stream to be set to `None`, which effectively disconnects it from the UART it would normally be connected to. It's returning zero for every write, which will also let the lwIP stack know that sending the packet failed and mark it as dropped, so it should be able to handle that correctly.

That is simpler and less error-prone than using a BytesIO object that might crash the board if it overflows. I've updated the example in the MR description to reflect this in the usage example.

DvdGiessen commented:
I noticed the new generic PPP class commits to the master branch, so updated this PR to also implement the changes there. This PR is now no longer specific to only the ESP32 port.

dpgeorge commented:
What sort of testing did you do against the generic `extmod/network_ppp_lwip.c` changes?

dpgeorge commented on docs/library/network.PPP.rst:
```diff
@@ -70,8 +70,11 @@ Methods
+   retrieved and set is the underlying stream, using::
+
+      stream = PPP.config("stream")
+      PPP.config("stream", stream)
```
Shouldn't this read `ppp.config(stream=stream)`?

DvdGiessen replied:
Indeed it should. 👍🏻

DvdGiessen commented:
> What sort of testing did you do against the generic extmod/network_ppp_lwip.c changes?

Not a lot; in fact I now noticed the commits I seemed to not have staged some changes and the generic extmod changes would not even compile. Sorry, I should have caught that earlier.

Pushed, and checked it by compiling for a STM32WB55 with lwIP enabled:

```
MicroPython v1.25.0-preview.37.g11b441b0b.dirty on 2024-11-06; NUCLEO-WB55 with STM32WB55RGV6
Type "help()" for more information.
>>> from machine import UART
>>> from network import PPP
>>> uart = UART(2)
>>> uart.irq().trigger()
0
>>> ppp = PPP(uart)
>>> uart.irq().trigger()
0
>>> ppp.connect()
>>> uart.irq().trigger()
16
>>> ppp.config('stream')
UART('LP1')
>>> ppp.config(stream=None)
>>> uart.irq().trigger()
0
>>> ppp.config('stream')
>>> ppp.config(stream=uart)
>>> uart.irq().trigger()
16
>>> ppp.config('stream')
UART('LP1')
```

Note we see the UART IRQ changing when it's set or unset as the PPP stream.

I don't have an non-ESP32 board with a modem here right now, so I haven't fully tested it with an active connection.

dpgeorge commented:
I have tested this PR on an stm32-based board (so using the `network_ppp_lwip.c` code):
- tested original PPP code that I had from when `network.PPP` was first developed, it works
- added a test case to do `config(stream=None)` then issue a HTTP GET request (which fails) then `config(stream=orig_stream)`, then issue an HTTPS GET request (which works)

So, I'm happy with the `network_ppp_lwip.c` functionality.

DvdGiessen commented:
Note I did test the ESP32-specific implementation extensively: Currently have it deployed in an application that has been running for over a year now and has been suspending the PPP connection every 15 minutes, retrieving some data via AT commands, and then resuming the PPP connection and successfully transmitting the collected data over it.

dpgeorge commented:
> Note I did test the ESP32-specific implementation extensively: Currently have it deployed in an application that has been running for over a year now

Is that application using the new scheme here that suspends the UART by setting it to `None`?  As long as the esp32-specific code as it stands now has been tested, then this PR is good to go.

DvdGiessen commented:
> Is that application using the new scheme here that suspends the UART by setting it to `None`? As long as the esp32-specific code as it stands now has been tested, then this PR is good to go.

Yes, it is using that new scheme since I implemented it back in February. Calling `ppp.config(stream=None)` to suspend, doing AT commands over the UART, and then and calling `self.ppp.config(stream=uart)` to resume.

dpgeorge commented:
> Yes, it is using that new scheme since I implemented it back in February

OK, great, thanks for confirming.
